src/rust/ide/lib/view/src/graph_editor/component/project_name.rs

```diff
@@ -0,0 +1,314 @@
+    /// Rename the project.
+    pub rename     : frp::Source<String>,
+    /// Reset the project name to the one before editing.
+    pub reset_name : frp::Source<()>
+}
+
+impl FrpInputs {
+    /// Create new FrpInputs.
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! {network
+            def reset_name = source();
+            def rename     = source();
+        }
+        Self{reset_name,rename}
+    }
+}
+
+
+
+// ==================
+// === FrpOutputs ===
+// ==================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct FrpOutputs {
+    /// Emits the new project name when it's renamed.
+    pub renamed : frp::Source<String>
+}
+
+impl FrpOutputs {
+    /// Create new FrpOutputs.
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! {network
+            def renamed = source();
+        }
+        Self{renamed}
+    }
+}
+
+// ===========
+// === Frp ===
+// ===========
+
+#[derive(Debug,Clone,CloneRef,Shrinkwrap)]
+#[allow(missing_docs)]
+pub struct Frp {
+    #[shrinkwrap(main_field)]
+    pub inputs  : FrpInputs,
+    pub outputs : FrpOutputs,
+    pub network : frp::Network,
+}
+
+impl Default for Frp {
+    fn default() -> Self {
+        let network = frp::Network::new();
+        let inputs  = FrpInputs::new(&network);
+        let outputs = FrpOutputs::new(&network);
+        Self{network,inputs,outputs}
+    }
+}
+
+impl Frp {
+    /// Create new Frp.
+    pub fn new() -> Self {
+        default()
+    }
+}
+
+
+
+// ==================
+// === Animations ===
+// ==================
+
+/// ProjectName's animations handlers.
+#[derive(Debug,Clone,CloneRef)]
+pub struct Animations {
+    highlight   : Animation<f32>,
+    positioning : Animation<Vector3<f32>>
+}
+
+impl Animations {
+    /// Create new animations handlers.
+    pub fn new(network:&frp::Network) -> Self {
+        let highlight   = Animation::<_>::new(&network);
+        let positioning = Animation::<_>::new(&network);
+        Self{highlight,positioning}
+    }
+}
+
+
+
+// ========================
+// === ProjectNameModel ===
+// ========================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct ProjectNameModel {
+    logger         : Logger,
+    animations     : Animations,
+    view           : component::ShapeView<shape::Shape>,
+    display_object : display::object::Instance,
+    text_field     : TextField,
+    project_name   : Rc<RefCell<String>>,
+    renamed_output : frp::Source<String>,
+}
+
+impl ProjectNameModel {
+    /// Create new ProjectNameModel.
+    pub fn new(world:&World,frp:&Frp) -> Self {
+        let scene                 = world.scene();
+        let logger                = Logger::new("ProjectName");
+        let display_object        = display::object::Instance::new(&logger);
+        let font                  = scene.fonts.get_or_load_embedded_font("DejaVuSansMono").unwrap();
+        let size                  = Vector2::new(600.0,100.0);
```

Comment:
hardocded constants here and there - moveall ofthem to section constants on top of the file